src/wallet/mod.rs

```diff
@@ -445,15 +446,53 @@ where
         self.database.borrow().iter_txs(include_raw)
     }
 
-    /// Return the balance, meaning the sum of this wallet's unspent outputs' values
+    /// Return the balance, separated into available, trusted-pending, untrusted-pending and immature
+    /// values.
     ///
     /// Note that this methods only operate on the internal database, which first needs to be
     /// [`Wallet::sync`] manually.
-    pub fn get_balance(&self) -> Result<u64, Error> {
-        Ok(self
-            .list_unspent()?
-            .iter()
-            .fold(0, |sum, i| sum + i.txout.value))
+    pub fn get_balance(&self) -> Result<Balance, Error> {
+        let mut immature = 0;
+        let mut trusted_pending = 0;
+        let mut untrusted_pending = 0;
+        let mut available = 0;
+        let utxos = self.list_unspent()?;
+        let database = self.database.borrow();
+        for u in utxos {
+            // Unwrap used since utxo set is created from database
+            let tx = database
+                .get_tx(&u.outpoint.txid, true)?
+                .expect("Transaction not found in database");
+            if let Some(tx_conf_time) = &tx.confirmation_time {
+                if tx.transaction.expect("No transaction").is_coin_base() {
+                    let last_sync_height = self
+                        .database()
+                        .get_sync_time()?
+                        .map(|sync_time| sync_time.block_time.height);
```

Comment:
This can be moved before the loop to avoid making extra unnecessary calls.

Moreover, I think you can just use `database` instead of `self.database()` since you've already borrowed it.

Reply:
Actually, even better: if you move this out of the loop then you can use a match and if the value is `None` you can just return `Balance::default()` because it means you've never synced in the first place.

This also lets you remove the `if let Some(last_sync_height) = last_sync_height` and remove a bit of the nesting